File: exercises/practice/circular-buffer/circular_buffer.zig
pub const BufferError = error{BufferOverflow};

pub fn CircularBuffer(comptime T: type, comptime capacity: usize) type {
    _ = capacity;
    return struct {
        const Self = @This();

        // This struct, as well as its fields and methods, needs to be implemented.

        /// Initializes a CircularBuffer
        pub fn init() Self {
            @compileError("please implement the init function");
        }

        /// Discards all items in the buffer.
        pub fn clear(self: *Self) void {
            _ = self;
            @compileError("please implement the clear function");
        }

        /// Extracts the oldest item from the buffer.
        pub fn read(self: *Self) ?T {
            _ = self;
            @compileError("please implement the read function");
        }

        /// Write `item` into the buffer.
        pub fn write(self: *Self, item: T) BufferError!void {
            _ = self;
            _ = item;
            @compileError("please implement the write function");
        }

        /// Write `item` into the buffer, replacing the oldest item if necessary.
        pub fn overwrite(self: *Self, item: T) void {
            _ = self;
            _ = item;
            @compileError("please implement the overwrite function");
        }
    };
}

File: generators/exercises/circular_buffer.py
HEADER = "const CircularBuffer = circular_buffer.CircularBuffer;\nconst BufferError = circular_buffer.BufferError;\n"


def gen_case(case):
    capacity = case["input"]["capacity"]
    operations = case["input"]["operations"]

    lines = [f"    var cb = CircularBuffer(i16, {capacity}).init();\n"]
    for op in operations:
        kind = op["operation"]
        if kind == "write":
            if op["should_succeed"]:
                lines.append(f"    try cb.write({op['item']});\n")
            else:
                lines.append(
                    f"    try testing.expectError(BufferError.BufferOverflow, cb.write({op['item']}));\n"
                )
        elif kind == "read":
            if op["should_succeed"]:
                lines.append(
                    f"    try testing.expectEqual({op['expected']}, cb.read());\n"
                )
            else:
                lines.append("    try testing.expectEqual(null, cb.read());\n")
        elif kind == "overwrite":
            lines.append(f"    cb.overwrite({op['item']});\n")
        elif kind == "clear":
            lines.append("    cb.clear();\n")
    return "".join(lines)
